numpy/core/src/multiarray/datetime.c

```diff
@@ -462,8 +462,21 @@ convert_datetime_to_datetimestruct(PyArray_DatetimeMetaData *meta,
             break;
 
         case NPY_FR_W:
-            /* A week is 7 days */
-            set_datetimestruct_days(dt * 7, out);
+            /*
+             * A week is 7 days but we can overflow
+             * if we apply the product directly, so
+             * iteratively account for week when near
+             * max int64 value
+             */
+            if ((dt > 0 && (dt > (NPY_MAX_INT64 / 7))) ||
+                (dt < 0 && (dt < NPY_MIN_INT64 / 7))) {
+                for (int i = 0; i < 7; ++i) {
+                    set_datetimestruct_days(dt, out);
```

Comment:
This doesn't make any sense to me - `set_datetimestruct_days` is a pure function that sets `out` based only on `dt`. Why are we calling a pure function in a loop with the same arguments each time?

Reply:
`dt` must be converted to the integer number of days, but is provided as the number of weeks. Looping 7 times is actually a very obvious way to break a week down into 7 calls for days.

The previous code was clearly broken so seems like big improvement to me. You can't just multiply an int64 by 7 and feed it to a "pure function"--the danger of overflow is clear with the old code.

This was discussed with Stephan above -- yes, I could probably use modulus to reduce the number of loops, but a week is seven days and that's a natural way to do this.

Reply:
I think I'm missing something here. Your approach makes sense if the function is `add_datetimestruct_days`. Does the function increment the number of days, or just set them?

I'm reading that loop as
```
for(...) {
    out = f(dt);
}
```
In which the loop doesn't do anything

Reply:
Maybe the function should be changed to an accumulator if it isn't already doing that, or an accumulating equivalent might be added eventually. I doubt we're doing any worse by avoiding undefined behavior from overflow and moving away from the epoch in at least the correct direction even if only at 1/7 of the desired magnitude in the worst case.

The most helpful thing is unit tests, and perhpas the need for gold standard reprs I mentioned. If I can't even write unit tests for the true limits because we can't generate the reprs programmatically, it is a bit crazy to sanity-check for correctness.

But as far as I can tell here, we never had the ability to properly accumulate days when weeks was on the order of 1/7 NPY_MAX_INT anyway.

Reply:
As written right now, it would be better to just set the output to NaT rather than produce a value off by a factor of 7.

Reply:
`set_datetimestruct_days` is still overflowing the years even with the loop because it calls `days_to_yearsdays`, which subtracts from `days` without an overflow guard, so we can't even safely  get to 1/7 of the span in negative direction at the moment.

Reply:
Can you link to those lines? Surely if  days is negative the subtraction results in an addition?

Reply:
Looks like a magic number subtraction to me here, with no special handling: https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/datetime.c#L159

Do you agree?

Reply:
Yep, that looks like a pain to resolve. Good catch.

Reply:
Either way - let's not introduce new code that is known to never produce correct results. If you don't want to deal with overflow here, just use `out->year = NPY_DATETIME_NAT;`, and stick a TODO on it